File: java/shared/java/org/signal/zkgroup/profiles/PniCredentialPresentation.java
//
// Copyright 2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.zkgroup.profiles;

import org.signal.zkgroup.InvalidInputException;
import org.signal.zkgroup.groups.ProfileKeyCiphertext;
import org.signal.zkgroup.groups.UuidCiphertext;
import org.signal.zkgroup.internal.ByteArray;
import org.signal.client.internal.Native;

public final class PniCredentialPresentation extends ByteArray {
  public PniCredentialPresentation(byte[] contents) throws InvalidInputException {
    super(contents);
    try {
      Native.PniCredentialPresentation_CheckValidContents(contents);
    } catch (IllegalArgumentException e) {
      throw new InvalidInputException(e.getMessage());
    }
  }

  public UuidCiphertext getAciCiphertext() {
    byte[] newContents = Native.PniCredentialPresentation_GetAciCiphertext(contents);

    try {
      return new UuidCiphertext(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public UuidCiphertext getPniCiphertext() {
    byte[] newContents = Native.PniCredentialPresentation_GetPniCiphertext(contents);

    try {
      return new UuidCiphertext(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public ProfileKeyCiphertext getProfileKeyCiphertext() {
    byte[] newContents = Native.PniCredentialPresentation_GetProfileKeyCiphertext(contents);

    try {
      return new ProfileKeyCiphertext(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

}

File: swift/Sources/LibSignalClient/zkgroup/PniCredentialPresentation.swift
//
// Copyright 2020-2022 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import SignalFfi

public class PniCredentialPresentation: ByteArray {

  public required init(contents: [UInt8]) throws {
    try super.init(contents, checkValid: signal_pni_credential_presentation_check_valid_contents)
  }

  public func getAciCiphertext() throws -> UuidCiphertext {
    return try withUnsafeBorrowedBuffer { buffer in
      try invokeFnReturningSerialized {
        signal_pni_credential_presentation_get_aci_ciphertext($0, buffer)
      }
    }
  }

  public func getPniCiphertext() throws -> UuidCiphertext {
    return try withUnsafeBorrowedBuffer { buffer in
      try invokeFnReturningSerialized {
        signal_pni_credential_presentation_get_pni_ciphertext($0, buffer)
      }
    }
  }

  public func getProfileKeyCiphertext() throws -> ProfileKeyCiphertext {
    return try withUnsafeBorrowedBuffer { buffer in
      try invokeFnReturningSerialized {
        signal_pni_credential_presentation_get_profile_key_ciphertext($0, buffer)
      }
    }
  }

}
